python/lsst/daf/butler/registry/dimensions/query.py

```diff
@@ -0,0 +1,118 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+from __future__ import annotations
+
+__all__ = ["QueryDimensionRecordStorage"]
+
+from typing import Optional
+
+import sqlalchemy
+
+from ...core import DataCoordinate, Dimension, DimensionElement, DimensionRecord, Timespan
+from ...core.dimensions.schema import makeElementTableSpec
+from ...core.utils import NamedKeyDict
+from ..interfaces import Database, DimensionRecordStorage
+from ..queries import QueryBuilder
+
+
+class QueryDimensionRecordStorage(DimensionRecordStorage):
+    """A read-only record storage implementation that aggregates the values of
+    another element for which it is a required or implied dimension.
+
+    Parameters
+    ----------
+    db : `Database`
+        Interface to the database engine namespace that will hold these
```

Comment:
I think I meant to say "engine and namespace" instead of "engine namespace". I am trying hard to use "namespace" instead of "schema" because the latter is even more overloaded.